### DISCUSSION TITLE
Performance of packing problems

### DISCUSSION THREAD

**Discussion by ktnr** [google/or-tools#3177] [CP-SAT questions]
I have observed considerable performance improvements for 2D packing problems since 9.2. That's great, thanks! I have compared the performance against the results reported in CJCM08 and obtained surprising results on the only two hard instances therein: E00N23, E00X23.

CJCM08 uses the following constraints/propagations, most of which have an equivalent in ortools:
1. Preprocessing techniques of CCM07
2. NoOverlap2D
3. Cumulative constraints (AddCumulative())
4. Propagations for cumulative constraints (use_cumulative_in_no_overlap_2d)
5. Disjunctive constraints in cumulative constraints (use_disjunctive_constraint_in_cumulative_constraint)
6. Time table edge finding (use_timetable_edge_finding_in_cumulative_constraint)
7. Energetic reasoning (use_overload_checker_in_cumulative_constraint)
8. Pruning the search tree by solving subset sum problems (no equivalent in ortools!?)

The "Basic" row uses: 1-6. The "All" row uses all of the above. The "Default" row uses ortools default parameters. The runtimes are: 

| Variant | E00N23 | E00X23 | 
| ------- | ------ | ------ |
| CJCM08-Basic  | 70s | 160s |
| CJCM08-All | 1s | 6s |
| ortools-Default | 182s | 5046s |
| ortools-Basic | 1676s | >14400s
| ortools-All| 468s | >14400s |

Note that the numbers in the CJCM08 rows are taken directly from the tables in the paper, i.e., run with a single thread on hardware from 2008. Unfortunately, the CP solver that was used is not specified in the paper. The ortools numbers are obtained with 8 threads on current hardware with version 9.3.10459. Here are the proto models: 
- [E00N23.txt](https://github.com/google/or-tools/files/8259901/E00N23.txt)
- [E00X23.txt](https://github.com/google/or-tools/files/8259903/E00X23.txt)

Alternatively, you can run the comparison with different parameters yourself using [this repo](https://github.com/ktnr/BinPacking2D/blob/master/experiments/ComparisonCJCM.py).

Do you have an intuition what might explain that performance difference?

CJCM08 also highlight the effectiveness of another solution method with sweep propagations by BCT06 for feasible instances. But, from Table 3 it seems to be less effective (maybe even counterproductive?) for infeasible instances. 
Does ortools use something like sweep propagations? If so, can it be disabled via a parameter? There are some references to "sweep" in resource.cc, diffn_util.cc, timetable.h/cc.

- [CJCM08] Clautiaux, F., Jouglet, A., Carlier, J., & Moukrim, A. (2008). A new constraint programming approach for the orthogonal packing problem. Computers & Operations Research, 35(3), 944-959.
- [CCM07] Clautiaux, F., Carlier, J., & Moukrim, A. (2007). A new exact method for the two-dimensional orthogonal packing problem. European Journal of Operational Research, 183(3), 1196-1211.
- [BCT06] Beldiceanu, N., Carlsson, M., & Thiel, S. (2006). Sweep synchronization as a global propagation mechanism. Computers & operations research, 33(10), 2835-2851.

**Comment by lperron** [comment]
Actually, the no_overlap_2d propagator is poor.
I want to implement stronger versions, but I lack the time to do that.

**Reply by ktnr** [reply]
I'm looking forward to when you do find the time and will rerun the comparisons then.

**Comment by ktnr** [comment]
Update for 9.4.1874

A lot of work has been done on scheduling problems. Among others, the new parameter use_energetic_reasoning_in_no_overlap_2d is available and use_cumulative_in_no_overlap_2d has been renamed to use_timetabling_in_no_overlap_2d  in 1605046dbaad05c2c9ffac2eb34daaaa5d8d2777, so I'll update the list from above:
1. Preprocessing techniques of CCM07
2. NoOverlap2D
3. Cumulative constraints (AddCumulative())
4. Disjunctive constraints in cumulative constraints (use_disjunctive_constraint_in_cumulative)
5. Time table edge finding (use_timetabling_in_no_overlap_2d)
6. Time table edge finding (use_timetable_edge_finding_in_cumulative_constraint)
7. Energetic reasoning (use_overload_checker_in_cumulative)
8. Energetic reasoning (use_energetic_reasoning_in_no_overlap_2d)
9. Pruning the search tree by solving subset sum problems (no equivalent in ortools!?)

I tested the following variants [here](https://github.com/ktnr/BinPacking2D/blob/730a1e6409023eea557640bbbb6ca64803d14bd5/experiments/ComparisonCJCM.py):
- Default: 1-4
- Default-V1: 1-4, 8
- Default-V2: 1-4, 5, 8
- Default-V3: 1-4, 5, 8, use_dual_scheduling_heuristics = False
- Default-V4: 1-4, use_dual_scheduling_heuristics = False
- Basic: 1-6
- Full: 1-8

| Variant | E00N23 | E00X23 | 
| ------- | ------ | ------ |
| CJCM08-Basic  | 70s | 160s |
| CJCM08-All | 1s | 6s |
| ortools-Default | 123s | 5010s |
| ortools-Default-V1 | 100s | 2812s |
| ortools-Default-V2 | 75s | 6018s |
| ortools-Default-V3 | 58s | 2761s |
| ortools-Default-V4 | 54s | 2525s |
| ortools-Basic | 643s | 10915s |
| ortools-All | 480s | 11392s |

For these two instances, it looks like a great improvement over 9.3. In ba15d312ba4cf5ed1cf802a83aec23cdd1c17aca, `use_dual_scheduling_heuristics` was enabled by default, so I disabled it in two variants. Note that #3407 is still unresolved at the time of testing. More hard instances for testing can be found [here](https://github.com/ktnr/BinPacking2D/tree/master/data/input/OPP/BPP-Subproblems).

Is the Diophantine solver from 0b0f33b8ce8c93bfb649945b38dfd1e8c33349af (not used in the tests as it came after the release of 9.4.1874) similar to a subset sum solver? Looks like it's somewhat related.

**Comment by lperron** [comment]
Thanks.

1) are you running with multiple workers ?
2) the diophantine presolve simplifies equations like bin_num_1 * var1 +
big_num_2 * var2 == small_cst
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00



Le ven. 19 août 2022 à 13:01, Leopold Kuttner ***@***.***> a
écrit :

**Reply by ktnr** [reply]
Yes, 8 Threads.

**Reply by lperron** [reply]
if you have the hardware, you can try 16

**Comment by lperron** [comment]
On my workstation (2018), with 16 workers (default settings), I solve E00N23 in 23s, and E00X23 in 630s. Still far from your code.

**Comment by ktnr** [comment]
Update for 9.5.2237

| Variant | E00N23 | E00X23 | 
| ------- | ------ | ------ |
| ortools-Default | 70s | 4742s |
| ortools-Default-V4 | 60s | 4805s |
| ortools-Basic | 287s | 10826s |
| ortools-All | 287s | 4298s |

**Comment by lperron** [comment]
Thanks for following up.
